### PR TITLE
fix(retainer): change mnesia table storage types during update

### PR DESCRIPTION
**Pull request by thalesmg**
https://emqx.atlassian.net/browse/EMQX-8650

### CODE REVIEW

File: changes/v5.0.14-zh.md [hunk 1]
```diff
@@ -14,3 +14,5 @@
 - 修复了测试GCP PubSub可能泄露内存的问题，以及其JWT令牌第二次刷新失败的问题。 [#9640](https://github.com/emqx/emqx/pull/9640)
 
 - 修复 MySQL 驱动断开连接时出现的数据丢失和匹配错误的问题 [#9638](https://github.com/emqx/emqx/pull/9638)。
+
+- 修正了一个问题，即如果不重新启动节点，改变内置数据库保留器的存储类型将不会生效 [#9676](https://github.com/emqx/emqx/pull/9676)。
```
Comment (lafirest) [inline changes/v5.0.14-zh.md:18]: - 修复了如果不重新启动节点，改变保留消息的存储类型将不会生效的问题 [#9676](https://github.com/emqx/emqx/pull/9676)。

**Review by lafirest — DISMISSED** [review]

File: apps/emqx_retainer/src/emqx_retainer.erl [hunk 2]
```diff
@@ -321,24 +321,31 @@ update_config(
...
+        true ->
             State#{
                 clear_timer := check_timer(
                     ClearTimer,
                     ClearInterval,
                     clear_expired
                 )
             };
-        _ ->
+        false ->
```
Comment (ieQu1) [inline apps/emqx_retainer/src/emqx_retainer.erl:348]: Looks like all the retained messages will be dropped, or?

Comment (thalesmg) [reply apps/emqx_retainer/src/emqx_retainer.erl:348]: Apparently, they are not lost.  Tested manually and they persist after the config change, and after a VM restart (if `disc`).  I've added a test assertion for that: https://github.com/emqx/emqx/pull/9676/commits/3c21cb442e775a21e9b2ec35eaa61cd424129148

File: apps/emqx_retainer/test/emqx_retainer_api_SUITE.erl [hunk 3]
```diff
@@ -219,6 +221,93 @@ t_lookup_and_delete(_) ->
 
     ok = emqtt:disconnect(C1).
 
+t_change_storage_type(_Config) ->
+    Path = api_path(["mqtt", "retainer"]),
+    {ok, ConfJson} = request_api(get, Path),
+    RawConf = emqx_json:decode(ConfJson, [return_maps]),
+    ct:print(">>>>>>>>>>>>> ~p", [RawConf]),
```
Comment (lafirest) [inline apps/emqx_retainer/test/emqx_retainer_api_SUITE.erl:228]: delete this line

Comment (thalesmg) [reply apps/emqx_retainer/test/emqx_retainer_api_SUITE.erl]: :see_no_evil:

**Review by ieQu1 — APPROVED** [review]

**Pull request merged**